onnxruntime/core/mlas/lib/quantize.cpp

```diff
@@ -210,11 +210,217 @@ Return Value:
+    const auto ZeroPointVectorA = MlasBroadcastInt32x4((int32_t)ZeroPointA);
+    const auto ZeroPointVectorB = MlasBroadcastInt32x4((int32_t)ZeroPointB);
+    const auto ZeroPointVectorC = MlasBroadcastInt32x4((int32_t)ZeroPointC);
+    const auto MinimumValueVectorC = MlasBroadcastFloat32x4(float(MinimumValue - ZeroPointC));
+    const auto MaximumValueVectorC = MlasBroadcastFloat32x4(float(MaximumValue - ZeroPointC));
+
+    size_t N = std::max(LengthA, LengthB);
+    if (LengthA == 1) {
+        auto IntegerVectorA = MlasBroadcastInt32x4((int32_t)*InputA);
+        auto FloatVectorA = MlasDequantizeLinearVector(IntegerVectorA, ScaleVectorA, ZeroPointVectorA);
+        while (N >= 4) {
+            auto IntegerVectorB = MlasQuantizeLinearUnpackBytes<DataType>(MlasBroadcastInt32x4(*((const int32_t*)InputB)));
+            auto FloatVectorC = MlasAddFloat32x4(
+                    FloatVectorA,
+                    MlasDequantizeLinearVector(IntegerVectorB, ScaleVectorB, ZeroPointVectorB));
+            auto IntegerVectorC = MlasQuantizeLinearVector(FloatVectorC, ScaleVectorC,
+                    MinimumValueVectorC, MaximumValueVectorC, ZeroPointVectorC);
+            IntegerVectorC = MlasQuantizeLinearPackBytes<DataType>(IntegerVectorC);
+
+#if defined(MLAS_NEON64_INTRINSICS)
+            vst1q_lane_s32((int32_t*)OutputC, IntegerVectorC, 0);
+#else
+            *((int32_t*)OutputC) = _mm_cvtsi128_si32(IntegerVectorC);
+#endif
+
+            InputB += 4;
+            OutputC += 4;
+            N -= 4;
+        }
+    }
+    else if (LengthB == 1) {
+        auto IntegerVectorB = MlasBroadcastInt32x4((int32_t)*InputB);
+        auto FloatVectorB = MlasDequantizeLinearVector(IntegerVectorB, ScaleVectorB, ZeroPointVectorB);
+        while (N >= 4) {
+            auto IntegerVectorA = MlasQuantizeLinearUnpackBytes<DataType>(MlasBroadcastInt32x4(*((const int32_t*)InputA)));
+            auto FloatVectorC = MlasAddFloat32x4(
+                    MlasDequantizeLinearVector(IntegerVectorA, ScaleVectorA, ZeroPointVectorA),
+                    FloatVectorB);
+            auto IntegerVectorC = MlasQuantizeLinearVector(FloatVectorC, ScaleVectorC,
+                    MinimumValueVectorC, MaximumValueVectorC, ZeroPointVectorC);
+            IntegerVectorC = MlasQuantizeLinearPackBytes<DataType>(IntegerVectorC);
+
+#if defined(MLAS_NEON64_INTRINSICS)
+            vst1q_lane_s32((int32_t*)OutputC, IntegerVectorC, 0);
+#else
+            *((int32_t*)OutputC) = _mm_cvtsi128_si32(IntegerVectorC);
+#endif
+
+            InputA += 4;
+            OutputC += 4;
+            N -= 4;
+        }
+    }
+    else {
+        while (N >= 4) {
+            auto IntegerVectorA = MlasQuantizeLinearUnpackBytes<DataType>(MlasBroadcastInt32x4(*((const int32_t*)InputA)));
+            auto IntegerVectorB = MlasQuantizeLinearUnpackBytes<DataType>(MlasBroadcastInt32x4(*((const int32_t*)InputB)));
+            auto FloatVectorC = MlasAddFloat32x4(
```

Comment:
same as the C kernel, is there an opportunity here for a shared function?

Reply:
You are right. Let me use template to achieve no runtime if ().

Reply:
Fixed!